ulid/providers/microsecond.py

```diff
@@ -0,0 +1,54 @@
+"""
+    ulid/providers/microsecond
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Contains data provider that uses precise timestamp and most significant randomness bits.
+"""
+import os
+
+from .. import hints
+from . import base, time
+
+
+class Provider(base.Provider):
+    """
+    Provider that uses microsecond values for timestamp and most significant randomness bits.
+    """
+    def __init__(self, default: base.Provider):
+        self.default = default
+
+    def new(self) -> base.TimestampRandomnessBytes:
```

Comment:
TODO: Need to reason about the math in this function some more.

I'm currently taking two bytes (16 bits) of the 10 in the randomness to store the microseconds. This is too much. We only 10 bits to store the microseconds (max value is 1000, so 2^10 -> 1024 will be enough). The code here needs to be reworked to make sure we make sure of these additional 6 bits properly and they come from the prng.

Reply:
Something like this would do:
```python
random_prefix = (microseconds << 6 | int.from_bytes(os.urandom(1), sys.byteorder) & 0x3f).to_bytes(2, 'big')
randomness = random_prefix + os.urandom(8)
```

Reply:
Thanks. I haven't had time to come back to this and finish but I'll take a look once I can.

Reply:
Resolved in latest commit and updated tests to extract timestamp bits to validate.